<a href="https://colab.research.google.com/github/pebbleeee/Marine-Ecosystem-Mining-Association-Rule-Analysis-of-Australian-Reefs/blob/main/CSCE676_Project_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# [CSCE676_Project] - Data Mining Checkpoint 2
**Author:** Yash Patel  
**GitHub Repository:** https://github.com/pebbleeee/Marine-Ecosystem-Mining-Association-Rule-Analysis-of-Australian-Reefs
---

# Environment Setup EDA (To Use Cleaned Dataset)

In [1]:
#Used this cell in order to verify when we should stop skipping lines and make sure everything could aline properly within a pandas dataframe.

import pandas as pd
import numpy as np

# 1. DATA LOADING
# We skip 71 lines so that line 72 becomes the column headers.
path = "/content/IMOS_-_National_Reef_Monitoring_Network_Sub-Facility_-_Global_reef_fish_abundance_and_biomass.csv"

try:
    # Use skiprows=71 to land exactly on the FID header
    df = pd.read_csv(path, skiprows=71, low_memory=False)

    # Document WHY: Standardizing column names to lowercase/stripped
    # ensures that 'method' is found regardless of extra spaces in the CSV.
    df.columns = df.columns.str.strip().str.lower()

    if 'method' in df.columns:
        print("✅ Header alignment successful! Found 'method' column.")

        # 2. FILTERING
        # Now we can safely filter for Method 1 and 2
        df_clean = df[df['method'].isin([1, 2])].copy()
        print(f"✅ Filtered {len(df_clean)} valid survey rows.")

        # Display the first 5 rows to confirm it looks right
        display(df_clean.head())
    else:
        print("❌ Error: 'method' not found. Actual columns are:", df.columns.tolist())

except Exception as e:
    print(f"❌ Failed to load file: {e}")

✅ Header alignment successful! Found 'method' column.
✅ Filtered 212150 valid survey rows.


fid  survey_id    country  \
0  ep_m1_public_data.fid-20ccf8b0_19c25193193_7546  812327444  Australia   
1  ep_m1_public_data.fid-20ccf8b0_19c25193193_7563  812327444  Australia   
2  ep_m1_public_data.fid-20ccf8b0_19c25193193_7572  812327444  Australia   
3  ep_m1_public_data.fid-20ccf8b0_19c25193193_757f  812327143  Australia   
4  ep_m1_public_data.fid-20ccf8b0_19c25193193_758c  812327143  Australia   

              area               ecoregion                  realm   location  \
0  South Australia  South Australian Gulfs  Temperate Australasia  Encounter   
1  South Australia  South Australian Gulfs  Temperate Australasia  Encounter   
2  South Australia  South Australian Gulfs  Temperate Australasia  Encounter   
3  South Australia  South Australian Gulfs  Temperate Australasia  Encounter   
4  South Australia  South Australian Gulfs  Temperate Australasia  Encounter   

  site_code             site_name  latitude  ...    phylum           class  \
0     GSV56              The Bird    -35.76  ...  Chordata  Actinopterygii   
1     GSV56              The Bird    -35.76  ...  Chordata  Actinopterygii   
2     GSV56              The Bird    -35.76  ...  Chordata  Actinopterygii   
3     GSV26  2nd Valley Boat Shed    -35.51  ...  Chordata  Actinopterygii   
4     GSV26  2nd Valley Boat Shed    -35.51  ...  Chordata  Actinopterygii   

               order         family            species_name  \
0  Tetraodontiformes  Monacanthidae  Meuschenia hippocrepis   
1  Tetraodontiformes  Monacanthidae  Meuschenia hippocrepis   
2  Tetraodontiformes  Monacanthidae  Meuschenia hippocrepis   
3        Perciformes    Pempheridae  Pempheris multiradiata   
4        Perciformes    Pempheridae  Pempheris multiradiata   

           reporting_name  size_class  total   biomass  \
0  Meuschenia hippocrepis        15.0    1.0   95.0367   
1  Meuschenia hippocrepis        20.0    1.0  202.4397   
2  Meuschenia hippocrepis        25.0    1.0  305.0548   
3  Pempheris multiradiata        10.0    1.0   25.0615   
4  Pempheris multiradiata        12.5    1.0   49.1671   

                                           geom  
0  POINT (138.07000732421875 -35.7599983215332)  
1  POINT (138.07000732421875 -35.7599983215332)  
2  POINT (138.07000732421875 -35.7599983215332)  
3   POINT (138.2100067138672 -35.5099983215332)  
4   POINT (138.2100067138672 -35.5099983215332)  

[5 rows x 30 columns]

#(1) Project Scope

**Dataset Overview:**

* Selected Dataset: Reef Life Survey (RLS) Australia Ecosystem Subset (2016–2026).

* Scale and Complexity: The dataset contains over 100,000 observational rows and 10,707 unique surveys, encompassing a high-dimensional space of 1,684 distinct species.

* System Challenge: The primary engineering hurdle is an extreme matrix sparsity of 98.44%, necessitating optimized data structures and careful algorithmic parameter tuning.

**EDA Findings Recap:**

* Taxonomic Distribution: The ecosystem is heavily dominated by Actinopterygii (Ray-finned fishes), which accounts for over 700,000 observations, while classes like Elasmobranchii (Sharks) and Reptilia remain relatively rare.

* Environmental Context: Survey activity is heavily clustered at the 5-meter depth contour, providing a high-confidence baseline for shallow reef-flat interactions.

* Pivoting & Cleaning: A robust preprocessing pipeline was implemented to bypass 71 lines of portal-generated metadata and pivot "long" observational data into a "wide" binary market-basket format.

**Data Mining Techniques**

**Course Techniques:**

* Frequent Itemset Mining & Association Rule Discovery: Identifying recurring sets of species and extracting directional relationships using metrics like Support, Confidence, and Lift to identify ecological dependencies.

* FP-Growth Algorithm: A high-performance, non-candidate generating tree structure specifically chosen for this project to handle the extreme 98.4% sparsity of the 1,684-species matrix efficiently.

* Graph Mining (Centrality & PageRank): Implementing Betweenness Centrality and PageRank (Standard and Personalized) to mathematically define "Keystone" species that serve as the primary hubs of the Australian reef network.

* Community Detection: Using graph partitioning to identify distinct "ecological modules" or groups of species that interact more frequently with each other than with the rest of the reef system.

* Clustering (k-means & DBSCAN): Grouping high-dimensional survey sites based on species composition and latent environmental variables, such as depth and biomass, to define regional biodiversity clusters.

**External Techniques:**

* Graph Embeddings (e.g., Node2Vec): Transforming the 1,684-species co-occurrence graph into a low-dimensional vector space (latent space) to capture complex, non-linear relationships and similarities between species that traditional rules may miss.

* Graph Neural Networks (GNNs): Utilizing a Message Passing framework to perform node-level classification, such as predicting a species' conservation status or biomass category based on the properties of its ecological neighbors.

* Network Motif Analysis: Systematically identifying recurring sub-graph structures (e.g., "predatory triangles" or "symbiotic chains") to characterize the topological stability and functional architecture of the reef ecosystem.

#(2) Research Question Definition

**RQ1: Topological Keystone Identification (Course)**

 * Research Question: Which species emerge as central hubs when comparing PageRank Centrality to simple Node Degree in the Australian reef co-occurrence network, and how does their removal impact network connectivity?

 * Data Mining Task Type: Graph Mining / Network Centrality Analysis.

 * Relevant Algorithm(s): PageRank (Course) and Degree Centrality.

 * Evaluation Criteria: Centrality scores, change in network fragmentation (global efficiency), and interpretability against known trophic levels.

**RQ2: Conservation Indicator Rules (Course)**

 * Research Question: How do varying Support and Lift thresholds influence the discovery of "Indicator Rules" between rare megafauna (e.g., Elasmobranchii) and high-biomass reef communities?

 * Data Mining Task Type: Frequent Itemset Mining / Association Rule Mining.

 * Relevant Algorithm(s): FP-Growth (Course).

 * Evaluation Criteria: Support, Confidence, Lift, and taxonomic diversity of the resulting itemsets.

**RQ3: Ecological Fingerprinting (External)**

 * Research Question: Do Network Motifs (recurring 3-node sub-graph structures) reveal structural architectural patterns in healthy reefs that are missed by global network centrality metrics?

 * Data Mining Task Type: Network Motif Discovery (Sub-graph Pattern Mining).

 * Relevant Algorithm(s): FANMOD or ISMA (Sub-graph Isomorphism - External).

 * Evaluation Criteria: Motif frequency (Z-score), P-value significance, and comparison of motif profiles across different ecoregions.

#(3) Motivation and Feasibility

In [5]:
import pandas as pd
import numpy as np
import networkx as nx

# 1. PIVOT & BINARIZATION
# Why: Transform long observation data into a binary sparse matrix (survey x species).
basket = df_clean.pivot_table(index='survey_id',
                               columns='species_name',
                               values='total',
                               aggfunc='sum',
                               fill_value=0)

# Why: Use .map to binarize counts (1 for presence, 0 for absence).
basket_sets = basket.map(lambda x: 1 if x > 0 else 0)
print("✅ basket_sets matrix defined.")

# 2. PREPARING CONNECTIVITY DATA
# Why: Identify target classes (Sharks/Turtles) to check their connectivity in the network.
target_classes = ['Elasmobranchii', 'Reptilia']
target_species = df_clean[df_clean['class'].isin(target_classes)]['species_name'].unique()

# 3. SPECIES DEGREE DISTRIBUTION
# Why: RQ1 (PageRank) requires connected nodes; disconnected species won't show centrality.
co_occurrence_matrix = basket_sets.T @ basket_sets
np.fill_diagonal(co_occurrence_matrix.values, 0)
species_degrees = (co_occurrence_matrix > 0).sum(axis=1)

print("\n--- FEASIBILITY: GRAPH CONNECTIVITY (RQ1) ---")
print(f"✅ Total Species in Network: {len(species_degrees)}")
print(f"✅ Average Connectivity (Degree): {species_degrees.mean():.2f}")
print(f"✅ Max Connectivity: {species_degrees.max()} species")

target_degrees = species_degrees[species_degrees.index.isin(target_species)]
print(f"✅ Target Indicator Connectivity: Mean={target_degrees.mean():.2f}, Min={target_degrees.min()}")

# 4. SPARSITY & MOTIF POTENTIAL
# Why: RQ3 (Motifs) needs sufficient edge density to find sub-graph structures.
edges = (co_occurrence_matrix.values > 0).sum() / 2
possible_edges = (len(basket_sets.columns) * (len(basket_sets.columns) - 1)) / 2
graph_density = edges / possible_edges

print("\n--- FEASIBILITY: EXTERNAL MOTIF MINING (RQ3) ---")
print(f"✅ Graph Density: {graph_density:.6%}")
print(f"✅ Total Biological Edges: {int(edges)}")

# 5. PRELIMINARY LIFT CHECK
# Why: RQ2 needs enough co-occurrences for sharks/turtles to generate valid rules.
print("\n--- FEASIBILITY: ASSOCIATION RULES (RQ2) ---")
high_co_occurrence = co_occurrence_matrix.loc[target_species].max(axis=1)
valid_targets = (high_co_occurrence >= 5).sum()
print(f"✅ {valid_targets}/{len(target_species)} Indicator species meet link threshold.")

if valid_targets > 0:
    print("✅ CONCLUSION: All 3 Research Questions are feasible.")
else:
    print("❌ WARNING: Indicator classes may require lower support thresholds.")

✅ basket_sets matrix defined.

--- FEASIBILITY: GRAPH CONNECTIVITY (RQ1) ---
✅ Total Species in Network: 1295
✅ Average Connectivity (Degree): 267.87
✅ Max Connectivity: 1105 species
✅ Target Indicator Connectivity: Mean=127.04, Min=6

--- FEASIBILITY: EXTERNAL MOTIF MINING (RQ3) ---
✅ Graph Density: 20.701306%
✅ Total Biological Edges: 173449

--- FEASIBILITY: ASSOCIATION RULES (RQ2) ---
✅ 25/55 Indicator species meet link threshold.
✅ CONCLUSION: All 3 Research Questions are feasible.


**General Motivation:**

The initial EDA confirmed that while the reef dataset is extremely sparse (98.44%), it possesses a highly interconnected "small-world" architecture with a graph density of 20.7% and over 173,000 biological edges. This suggests that species do not exist in isolation; rather, they form a complex web where connectivity is a better measure of ecosystem health than simple abundance counts.

**RQ1: Topological Keystone Identification (Course: PageRank/Degree)**

 * Motivation: The EDA showed an average connectivity of 267.87, indicating a dense network of co-occurrences. Identifying which species act as the "mathematical hubs" allows us to prioritize conservation for species that anchor the network's stability.

 * Non-triviality: Standard biological metrics often focus on biomass. PageRank is non-trivial because it identifies "Hidden Keystones"—species that may be low in biomass but are topologically critical because they connect otherwise isolated ecological modules.

 * Feasibility: Target indicators (Sharks/Turtles) have a mean connectivity of 127.04. This high degree ensures that PageRank will have significant "signal" to rank these species within the global network.

 * Risks: Centrality results can be sensitive to the damping factor; results may be dominated by the most ubiquitous species if not properly normalized.

**RQ2: Conservation Indicator Rules (Course: FP-Growth)**

 * Motivation: The feasibility check identified that 25 out of 55 rare indicator species meet the minimum co-occurrence threshold for rule generation. This proves we can mine meaningful "indicator rules" even for rare, high-protection classes.

 * Non-triviality: While we know these species are rare, we do not know if their presence statistically predicts the presence of other healthy biodiversity. This requires association mining to move from simple observation to predictive modeling.

 * Feasibility: FP-Growth is highly feasible for this dataset as it efficiently handles the 1,295-species high-dimensional space without the candidate-generation bottleneck of the Apriori algorithm.

 * Risks: Setting the minimum support too low could lead to a combinatorial explosion of trivial rules; requires strict Lift filtering.

**RQ3: Ecological Fingerprinting (External: Network Motifs)**

 * Motivation: With over 173,449 edges, the graph provides a rich topological landscape. Global centrality (RQ1) misses the local "functional architecture" (e.g., three-way symbiotic loops) that define a healthy reef's DNA.

 * Non-triviality: Course techniques look at individual nodes or unordered sets. Motif analysis is an external technique that treats the reef as a structured system, looking for specific sub-graph isomorphisms that represent stable biological interactions.

 * Feasibility: The graph size (1,295 nodes) is within the computational limits for motif-finding libraries like NetworkX or FANMOD. The density is high enough to ensure motifs exist but not so high that the search space becomes infinite.

 * Risks: High computational cost of searching for motifs larger than 3 nodes; sensitive to noise in the initial co-occurrence matrix.


#(4) Methodological Planning

**RQ1: Topological Keystone Identification**

 * Course Algorithm: PageRank Centrality and Degree Centrality.

 * Baselines: Simple Species Abundance and Node Degree. We will compare if the "most common" species are actually the most "central" to the network.

 * Evaluation: Network Fragmentation (measuring the increase in disconnected components when top hubs are removed) and Global Efficiency.

**RQ2: Conservation Indicator Rules**

* Course Algorithm: FP-Growth.

* Baselines: Apriori Algorithm (to justify why FP-Growth is needed for high-dimensional sparsity) and High-Support-Only Mining.

* Evaluation: Lift (to measure strength of association) and Jaccard Similarity (to measure rule stability across different depth subsets).

**RQ3: Ecological Fingerprinting**

* External Algorithm: Sub-graph Isomorphism / Network Motif Discovery.

* Baselines: Random Graph (Erdős–Rényi). We will compare our reef network against a randomly generated graph of the same density to see which patterns are non-random.

* Evaluation: Z-score (significance of motif frequency) and P-value.

In [9]:
import pandas as pd
import numpy as np
import networkx as nx
from mlxtend.frequent_patterns import fpgrowth, association_rules
import warnings
warnings.filterwarnings("ignore", category=DeprecationWarning)

# --- INITIAL METHOD RUNS ---
# Why: Verify library compatibility and computational efficiency for all RQs.

# 1. RQ1 INITIAL RUN: PAGERANK (Course)
# Why: Test if the species graph can be ranked to find hubs.
print("--- RQ1 INITIAL RUN: PAGERANK ---")
G = nx.from_pandas_adjacency(co_occurrence_matrix)
pagerank_scores = nx.pagerank(G, alpha=0.85)
top_species = sorted(pagerank_scores.items(), key=lambda x: x[1], reverse=True)[:5]
print(f"✅ Top 5 Keystone Candidates: {top_species}")

# 2. RQ2 INITIAL RUN: FP-GROWTH (Course)
# Why: Test if the sparse matrix handles tree-generation without memory errors.
print("\n--- RQ2 INITIAL RUN: FP-GROWTH ---")
# Using 0.05 support to account for high sparsity
freq_items = fpgrowth(basket_sets, min_support=0.05, use_colnames=True)
rules = association_rules(freq_items, metric="lift", min_threshold=1.2)
print(f"✅ Generated {len(rules)} rules.")
print(f"✅ Max Lift found: {rules['lift'].max():.2f}")

# 3. RQ3 INITIAL RUN: MOTIF (External)
# Why: Test computational cost of finding 3-node subgraphs (triangles).
print("\n--- RQ3 INITIAL RUN: MOTIF (TRIANGLES) ---")
triangles = sum(nx.triangles(G).values()) / 3
print(f"✅ Detected {int(triangles)} closed triangular motifs in the reef network.")

print("\n--- FINAL FEASIBILITY STATUS ---")
print("✅ All packages (NetworkX, mlxtend) are compatible and efficient.")

--- RQ1 INITIAL RUN: PAGERANK ---
✅ Top 5 Keystone Candidates: [('Labroides dimidiatus', 0.01077366815663258), ('Chlorurus sordidus', 0.008704049756153789), ('Thalassoma lunare', 0.008661449218104402), ('Thalassoma lutescens', 0.007206428826468924), ('Ctenochaetus striatus', 0.006742930643800311)]

--- RQ2 INITIAL RUN: FP-GROWTH ---
✅ Generated 18250 rules.
✅ Max Lift found: 13.50

--- RQ3 INITIAL RUN: MOTIF (TRIANGLES) ---
✅ Detected 14683968 closed triangular motifs in the reef network.

--- FINAL FEASIBILITY STATUS ---
✅ All packages (NetworkX, mlxtend) are compatible and efficient.


#Collaboration Declaration


On my honor, I declare the following resources:

1.Collaborators:
 * None

2.Web Sources:

 * Dataset: Reef Life Survey (RLS) Australia Ecosystem Subset (2016–2026): https://portal.aodn.org.au/

 * Library Documentation: NetworkX Documentation for Graph Mining and Mlxtend for FP-Growth implementation.

3.AI Tools:

 * Gemini (AI Assistant): I collaborated extensively with Gemini to refine the project's technical scope and handle the extreme 98.44% sparsity of the Australian Reef Life Survey dataset. Through iterative prompting, I utilized the AI to design a robust preprocessing pipeline that bypassed 71 lines of portal-generated metadata and pivoted "long" observational data into a binary "market basket" format. Furthermore, the AI assisted in translating my interest in systems-level network analysis into concrete data mining tasks, including evaluating the feasibility of PageRank Centrality for identifying keystone species, implementing FP-Growth for association rule mining, and selecting Network Motif Analysis as a novel external technique. Finally, the AI helped troubleshoot package compatibility issues and internal deprecation warnings during the initial method runs, ensuring that the 1,295-node species graph was computationally efficient for motif discovery and centrality ranking.